c10/util/typeid.h

```diff
@@ -343,60 +341,78 @@ class C10_API TypeMeta final {
 
   TypeMeta(TypeMeta&& rhs) noexcept = default;
 
- private:
+private:
   // TypeMeta can only be created by Make, making sure that we do not
   // create incorrectly mixed up TypeMeta objects.
-  explicit TypeMeta(const detail::TypeMetaData* data) noexcept
-  : data_(data) {
+  explicit TypeMeta(const uint16_t index) noexcept
+  : index_(index) {
   }
 
  public:
   /**
    * Returns the type id.
    */
   TypeIdentifier id() const noexcept {
-    return data_->id_;
+    return data().id_;
+  }
+  /**
+   * true if we represent some ScalarType type
+   */
+  inline bool isScalarType() const noexcept {
+    return index_ < NumScalarTypes;
+  }
+  /**
+   * true if we represent ScalarType scalar_type
+   */
+  inline bool isScalarType(ScalarType scalar_type) const noexcept {
+    return index_ == static_cast<uint16_t>(scalar_type);
   }
   /**
    * Returns the size of the item.
    */
-  size_t itemsize() const noexcept {
-    return data_->itemsize_;
+  inline size_t itemsize() const noexcept {
+    if (C10_LIKELY(isScalarType())) {
+      return scalarTypeItemSizes[index_];
```

Comment:
Curious: what's the top line IR count impact of the fast path compared to using the general data().itemsize_ below? What should be the general guideline? e.g.: whether to create a fast path for, say, copy()?

Reply:
Ugh that's a good question and I don't have the callgrind data saved for this change - informally, anything that looked like it approached ~1% caught my eye - basically anything on the first page or so of callgrind listings that seemed like it was unreasonably high IR. 

itemsize() is called within empty_cpu(), which is why I saw it and none of the others in the benchmarks I was running, but it seemed like a decent balance to stop there, rather than do fast paths for the others on general principles. I figured if we see evidence that particular others are hot we can always add fast paths for them.

Reply:
I'm guessing if you reserve space in `data` for the itemsize table rather than indirecting through vector you can eliminate the necessity for this fastpath.

Reply:
Not sure if it's what you mean, but yeah: always getting itemsize from the TypeMetaData array entries is close enough now to demotivate the side table - only about 0.1% difference. Putting up an update with that simplification, but LMK if you meant something smarter.

Reply:
Interestingly, I see what appear to be consistently better numbers on our in-house microbenchmarks with the fast path `itemsize()` is left in. Noise makes the source of the difference hard to bisect, but the difference across the stack is notable. I think I'll leave the fast path in at least until we can use @robieta's stuff to bisect with a bit less manual labor :)